File: Programmers/2/150368. 이모티콘 할인행사/이모티콘 할인행사.java
class Solution {
    
    int signUp = 0;
    int revenue = 0;
    
    public int[] solution(int[][] users, int[] emoticons) {
        int[] answer = new int[2];

        int[] offPrice = new int[emoticons.length];
        
        applyDiscount(offPrice, 0, users, emoticons);
        
        answer[0] = signUp;
        answer[1] = revenue;
        
        return answer;
    }
    
    private void applyDiscount(int[] offPrice, int index, int[][] users, int[] emoticons) {
        if (index == offPrice.length) {
            calculateProfit(offPrice, users, emoticons);
            
            return;
        }
        
        for (int i = 10; i <= 40; i += 10) {
            offPrice[index] = i;
            
            applyDiscount(offPrice, index + 1, users, emoticons);
        }
    }
    
    private void calculateProfit(int[] offPrice, int[][] users, int[] emoticons) {
        int count = 0;
        int discountRevenue = 0;
        
        for (int[] user : users) {
            int discount = user[0];
            int price = user[1];
            int sum = 0;
            
            for (int i = 0; i < offPrice.length; i++) {
                if (offPrice[i] >= discount) {
                    sum += (emoticons[i] / 100) * (100 - offPrice[i]);
                }
            }
            
            if (sum >= price) {
                count++;
            } else {
                discountRevenue += sum;
            }
        }
        
        if (count > signUp) {
            signUp = count;
            revenue = discountRevenue;
            
            return;
        } else if (count == signUp) {
            if (revenue < discountRevenue) {
                revenue = discountRevenue;
            }
        }
    }
}

File: Programmers/2/150368. 이모티콘 할인행사/이모티콘 할인행사.cpp
#include <string>
#include <vector>

using namespace std;

const vector<int> DISCOUNTS = {10, 20, 30, 40};
int maxSubscribers = 0;
int maxRevenue = 0;

void simulate(const vector<int>& discounts,
             const vector<vector<int>>& users,
             const vector<int>& emoticons)
{
    int subscribers = 0;
    int revenue = 0;
    
    for (const auto& user : users)
    {
        int minDiscount = user[0];
        int maxPrice = user[1];
        int totalCost = 0;
        
        for (int i = 0; i < emoticons.size(); i++)
        {
            if (discounts[i] >= minDiscount)
            {
                totalCost += emoticons[i] * (100 - discounts[i]) / 100;
            }
        }
        
        if (totalCost >= maxPrice)
        {
            subscribers++;
        }
        else
        {
            revenue += totalCost;
        }
    }
    
    if (subscribers > maxSubscribers ||
       (subscribers == maxSubscribers && revenue > maxRevenue))
    {
        maxSubscribers = subscribers;
        maxRevenue = revenue;
    }
}

void dfs(int depth, vector<int>& discounts,
        const vector<vector<int>>& users,
        const vector<int>& emoticons)
{
    if (depth == emoticons.size())
    {
        simulate(discounts, users, emoticons);
        return;
    }
    
    for (int discount : DISCOUNTS)
    {
        discounts[depth] = discount;
        dfs(depth + 1, discounts, users, emoticons);
    }
}

vector<int> solution(vector<vector<int>> users, vector<int> emoticons) {
    vector<int> discounts(emoticons.size());
    
    dfs(0, discounts, users, emoticons);
    
    return {maxSubscribers, maxRevenue};
}